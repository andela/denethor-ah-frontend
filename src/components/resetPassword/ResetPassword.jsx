import React, { Component } from 'react';
import { isAlphanumeric } from 'validator';
import PropTypes from 'prop-types';
import '@babel/polyfill';
import { toast } from 'react-toastify';
import './styles.scss';
import resetPassword from './resetPasswordAction';

export class ResetPassword extends Component {
  state = {
    buttonDisabled: false,
    newPassword: '',
    oldPassword: '',
    newPasswordErrorMessage: '',
    oldPasswordErrorMessage: '',
    toastId: 'toast'
  }


  onOldPasswordChange = ({ target: { value: oldPassword } }) => {
    this.setState({
      oldPassword,
      oldPasswordErrorMessage: ''
    });
  }

  onNewPasswordChange = ({ target: { value: newPassword } }) => {
    this.setState({
      newPassword,
      newPasswordErrorMessage: '',
    });
  }

  onOldPasswordBlur = () => {
    const { oldPassword } = this.state
    if (!oldPassword) {
      return this.setState({ oldPasswordErrorMessage: 'Please enter your Old password.' })
    }
  }

  onNewPasswordBlur = () => {
  const { newPassword, oldPassword } = this.state

  if (!newPassword) {
    return this.setState({ newPasswordErrorMessage: 'Please enter your New password.' })
  }

  if (!isAlphanumeric(newPassword)) {
    return this.setState({ newPasswordErrorMessage: 'Use numbers and letters for password.' })
  }

  if (newPassword.length < 8) {
    return this.setState({ newPasswordErrorMessage: 'Password must be 8 or more characters.' })
  }

  if (oldPassword === newPassword) {
    return  this.setState({newPasswordErrorMessage: 'New password must be different.'})
  }
}

  handleSubmit = (e) => {
    e.preventDefault();
    const { toastId, oldPassword, newPassword } = this.state;
    if (!oldPassword) {
      return this.setState({ oldPasswordErrorMessage: 'Please enter your Old password.' });
    }
    if (!newPassword) {
      return this.setState({ newPasswordErrorMessage: 'Please enter your New password.' });
    }

    this.setState({ buttonDisabled: true });

    this.props.resetPassword(oldPassword, newPassword)
     .then(() => 
     toast.success('Password set successfully', {
      autoClose: true,
      hideProgressBar: true,
      toastId
    })
    )
     .catch(error => {
       const { response } = error;
       if (response && response.status === 422){
        return toast.error(response.data.data.input, {
          hideProgressBar: true,
          autoClose: true,
          toastId
        });
      }
      if(response && response.status === 403){
        return toast.error(response.data.message, {
          hideProgressBar: true,
          autoClose: true,
          toastId
        });
      }
      if (response && response.status === 500){
        return toast.error('Server error', {
        hideProgressBar: true,
        autoClose: true,
        toastId
      });
      }
      else{toast.error('unKnown error', {
        hideProgressBar: true,
        autoClose: true,
        toastId
      });
      }
      })
    this.setState({ buttonDisabled: false });

  }

  render() {
    const {
      buttonDisabled,
      newPassword,
      oldPasswordErrorMessage,
      oldPassword,
      newPasswordErrorMessage,
    } = this.state;

    const disableButton = buttonDisabled || oldPasswordErrorMessage || newPasswordErrorMessage;

    return (
      <div className='flex modal-wrapper'>
        <div className='modal-form-container'>
          <div className='modal-header'>
            <div className="logo-container">
              <h3>Enter Password</h3>
            </div>
          </div>
          <div className='reset-password-form__container'>
            <form className='reset-password-form__form'>
              <div>
                <small className='password-error-message'>
                  &nbsp;{oldPasswordErrorMessage}
                </small>
                <input
                  className={oldPasswordErrorMessage && 'red-bg'}
                  type='password'
                  name='password'
                  placeholder='Old Password'
                  value={oldPassword}
                  onChange={this.onOldPasswordChange}
                  onBlur={this.onOldPasswordBlur}
                  required
                />
              </div>

              <div>
                <small className='password-error-message'>
                  &nbsp;{newPasswordErrorMessage}
                </small>
                <input
                  className={newPasswordErrorMessage && 'red-bg'}
                  type='password'
                  name='confirm-password'
                  onBlur={this.onNewPasswordBlur}
                  placeholder='New Password'
                  value={newPassword}
                  onChange={this.onNewPasswordChange}
                  required
                />
              </div>
              <button
                disabled={disableButton}
                className='button button--primary'
                onClick={this.handleSubmit}
                name ='submit'
              >
                <span className={`button__text--login ${buttonDisabled ? 'hidden' : undefined}`}>
                  Submit
                </span>
                <span className={buttonDisabled ? 'spinner spinner--button' : undefined}></span>
              </button>
            </form>
          </div>
        </div>
      </div>
    );
  }
}
ResetPassword.propTypes = { 
  resetPassword: PropTypes.func
}
ResetPassword.defaultProps = {
  resetPassword
}
export default (ResetPassword);
