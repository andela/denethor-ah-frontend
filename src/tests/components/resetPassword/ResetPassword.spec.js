import React from 'react';
import { shallow } from 'enzyme';
import ResetPassword from '../../../components/resetPassword';
import axios from '../../../utils/axiosConfig';


jest.mock('../../../utils/axiosConfig');
jest.mock('react-toastify');
jest.mock('../../../redux/actions/auth');

let resetPassword;
let wrapper;
const preventDefault = jest.fn();

beforeEach(() => {
  wrapper = shallow(<ResetPassword />);
})

test('Should render ResetPassword correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('Should set state on password change', () => {
  wrapper.find('input').at(0).simulate('change', {
    target: { value: 'm' }
  });

  expect(wrapper.state('oldPassword')).toBe('m');

  wrapper.find('input').at(1).simulate('change', {
    target: { value: 'qwertyuiop' }
  });

  expect(wrapper.state('newPassword')).toBe('qwertyuiop');
});

test('Should verify on password blur', () => {
  wrapper.find('input').at(1).simulate('change', {
    target: { value: 'mon' }
  });
  wrapper.find('input').at(1).simulate('blur');

  expect(wrapper.state('newPasswordErrorMessage')).toBe('Password must be 8 or more characters.');

  wrapper.find('input').at(1).simulate('change', {
    target: { value: 'monkey&*&*' }
  });

  wrapper.find('input').at(1).simulate('blur');

  expect(wrapper.state('newPasswordErrorMessage')).toBe('Use numbers and letters for password.');

  wrapper.find('input').at(0).simulate('change', {
    target: { value: '' }
  });

  wrapper.find('input').at(0).simulate('blur');

  expect(wrapper.state('oldPasswordErrorMessage')).toBe('Please enter your Old password.');

  wrapper.find('input').at(1).simulate('change', {
    target: { value: '' }
  });

  wrapper.find('input').at(1).simulate('blur');

  expect(wrapper.state('newPasswordErrorMessage')).toBe('Please enter your New password.');

  wrapper.find('input').at(0).simulate('change', {
    target: { value : 'olubukola'}
  })

  wrapper.find('input').at(1).simulate('change', {
  target: { value: 'olubukola'}
  })

  wrapper.find('input').at(1).simulate('blur');

  expect(wrapper.state('newPasswordErrorMessage')).toBe('New password must be different.');
});

test('Should throw error if submit is clicked when old Password is not empty', async () => {

  wrapper.find('button[name="submit"]').simulate('click', { preventDefault });
  expect(preventDefault).toHaveBeenCalled();
  expect(wrapper.state('oldPasswordErrorMessage')).toBe('Please enter your Old password.');
});

test('Should throw error if submit is clicked when new Password is not empty', async () => {
  wrapper.find('input').at(0).simulate('change', {
    target: { value: 'qwertyuiop'}
  })

  wrapper.find('button[name="submit"]').simulate('click', { preventDefault });
  expect(preventDefault).toHaveBeenCalled();
  expect(wrapper.state('oldPassword')).toBe('qwertyuiop');
  expect(wrapper.state('newPasswordErrorMessage')).toBe('Please enter your New password.');
});

test('Should submit successfully if correct inputs is provided', async () => {

  const props = {
    resetPassword: jest.fn().mockImplementation(() => Promise.resolve({}))
  };

  wrapper = shallow(<ResetPassword  {...props} />);

  wrapper.find('input').at(0).simulate('change', {
    target: { value: 'qwertyuiop' }
  });

  wrapper.find('input').at(1).simulate('change', {
    target: { value: 'qwertyuiop' }
  });


  wrapper.find('button').simulate('click', { preventDefault });
  expect(props.resetPassword).toHaveBeenCalled();

});
test('Should throw error if inputs are invalid', async () => {

  wrapper.find('input').at(0).simulate('change', {
    target: { value: 'qwertyuiop' }
  });

  wrapper.find('input').at(1).simulate('change', {
    target: { value: 'qwertyuiop' }
  });

  wrapper.find('button').simulate('click', { preventDefault });

  resetPassword = jest.fn(() => Promise.reject({}));
  axios.patch.mockResolvedValue(resetPassword);
  wrapper.find('button').simulate('click', { preventDefault });
  
});

it('should show a toast error when user Old password does not exist', () => {
 const resetPassword = jest.fn().mockImplementation(() => Promise.reject({response:{
    status:403
  }}))
  const props = {
    resetPassword
  };

  wrapper = shallow(<ResetPassword  {...props} />);
  wrapper.find('input').at(0).simulate('change', {
    target: { value: 'qwertyuiop' }
  });

  wrapper.find('input').at(1).simulate('change', {
    target: { value: 'qwertyuiop' }
  });


  wrapper.find('button').simulate('click', { preventDefault });

});

it('should show a toast error when a validation error occurs', () => {
  const resetPassword = jest.fn().mockImplementation(() => Promise.reject({response:{
     status:422
   }}))
   const props = {
     resetPassword
   };
 
   wrapper = shallow(<ResetPassword  {...props} />);
   wrapper.find('input').at(0).simulate('change', {
     target: { value: 'qwertyuiop' }
   });
 
   wrapper.find('input').at(1).simulate('change', {
     target: { value: 'qwertyuiop' }
   });
 
   wrapper.find('button').simulate('click', { preventDefault });
 
 });