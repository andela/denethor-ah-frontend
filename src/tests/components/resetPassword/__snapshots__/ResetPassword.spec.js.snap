// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ResetPassword correctly 1`] = `
<div
  className="flex modal-wrapper"
>
  <div
    className="modal-form-container"
  >
    <div
      className="modal-header"
    >
      <div
        className="logo-container"
      >
        <h3>
          Enter Password
        </h3>
      </div>
    </div>
    <div
      className="reset-password-form__container"
    >
      <form
        className="reset-password-form__form"
      >
        <div>
          <small
            className="password-error-message"
          >
             
          </small>
          <input
            className=""
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="Old Password"
            required={true}
            type="password"
            value=""
          />
        </div>
        <div>
          <small
            className="password-error-message"
          >
             
          </small>
          <input
            className=""
            name="confirm-password"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="New Password"
            required={true}
            type="password"
            value=""
          />
        </div>
        <button
          className="button button--primary"
          disabled=""
          name="submit"
          onClick={[Function]}
        >
          <span
            className="button__text--login undefined"
          >
            Submit
          </span>
          <span />
        </button>
      </form>
    </div>
  </div>
</div>
`;
